File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/config/ClientConfiguration.kt
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */

package software.amazon.smithy.swift.codegen.config

import software.amazon.smithy.codegen.core.Symbol
import software.amazon.smithy.swift.codegen.Dependency
import software.amazon.smithy.swift.codegen.integration.ProtocolGenerator
import software.amazon.smithy.swift.codegen.model.buildSymbol

/**
 * Specifies the behaviour of the service configuration
 */
interface ClientConfiguration {
    /**
     * The protocol name of the client configuration
     */
    val swiftProtocolName: Symbol?

    fun getProperties(ctx: ProtocolGenerator.GenerationContext): Set<ConfigProperty>

    companion object {
        fun runtimeSymbol(
            name: String,
            dependency: Dependency?,
        ): Symbol = buildSymbol {
            this.name = name
            dependency?.also {
                this.namespace = it.target
                dependency(dependency)
            }
        }
    }
}

File: Sources/ClientRuntime/Config/ClientConfiguration.swift
//
// Copyright Amazon.com Inc. or its affiliates.
// All Rights Reserved.
//
// SPDX-License-Identifier: Apache-2.0
//

/// Base protocol for client configuration.
///
/// Note: This protocol will inherit `Sendable` after the deprecation period of class-based configurations.
/// The struct-based configuration already conforms to `Sendable` via `@unchecked Sendable`.
public protocol ClientConfiguration {
    init() async throws
}
